frontend/device/kobo/device.lua

```diff
@@ -1238,6 +1238,117 @@ function Kobo:isStartupScriptUpToDate()
+        UIManager.event_handlers["SleepCoverClosed"] = function()
+            UIManager.event_handlers["Suspend"]()
+        end
+        UIManager.event_handlers["SleepCoverOpened"] = function()
+            self.is_cover_closed = false
+        end
+    else
+        UIManager.event_handlers["SleepCoverClosed"] = function()
+            self.is_cover_closed = true
+            UIManager.event_handlers["Suspend"]()
+        end
+        UIManager.event_handlers["SleepCoverOpened"] = function()
+            self.is_cover_closed = false
+            UIManager.event_handlers["Resume"]()
+        end
+    end
+    UIManager.event_handlers["Light"] = function()
+        self:getPowerDevice():toggleFrontlight()
+    end
+    -- USB plug events with a power-only charger
+    UIManager.event_handlers["Charging"] = function()
+        self:_beforeCharging()
+        -- NOTE: Plug/unplug events will wake the device up, which is why we put it back to sleep.
+        if self.screen_saver_mode then
+           UIManager.event_handlers["Suspend"]()
+        end
+    end
+    UIManager.event_handlers["NotCharging"] = function()
+        -- We need to put the device into suspension, other things need to be done before it.
+        self:usbPlugOut()
+        self:_afterNotCharging()
+        if self.screen_saver_mode then
+           UIManager.event_handlers["Suspend"]()
+        end
+    end
+    -- USB plug events with a data-aware host
+    UIManager.event_handlers["UsbPlugIn"] = function()
+        self:_beforeCharging()
+        -- NOTE: Plug/unplug events will wake the device up, which is why we put it back to sleep.
+        if self.screen_saver_mode then
+            UIManager.event_handlers["Suspend"]()
+        else
+            -- Potentially start an USBMS session
+            local MassStorage = require("ui/elements/mass_storage")
+            MassStorage:start()
+        end
+    end
+    UIManager.event_handlers["UsbPlugOut"] = function()
+        -- We need to put the device into suspension, other things need to be done before it.
+        self:usbPlugOut()
+        self:_afterNotCharging()
+        if self.screen_saver_mode then
+            UIManager.event_handlers["Suspend"]()
+        else
+            -- Potentially dismiss the USBMS ConfirmBox
+            local MassStorage = require("ui/elements/mass_storage")
+            MassStorage:dismiss()
+        end
+    end
+    UIManager.event_handlers["__default__"] = function(input_event)
+        -- Suspension in Kobo can be interrupted by screen updates. We ignore user touch input
+        -- in screen_saver_mode so screen updates won't be triggered in suspend mode.
+        -- We should not call self:suspend() in screen_saver_mode lest we stay on forever
+        -- trying to reschedule suspend. Other systems take care of unintended wake-up.
+        if not self.screen_saver_mode then
+            UIManager:sendEvent(input_event)
+        end
+    end
+end
```

Comment:
I will look into this tomorrow, and yes cervantes is the same except the cover stuff. The others are also be no problem (except I cannot test them).

Now the question arose: There is a bit of inconstancy between the usage of the onSuspend, onReboot and onPowerOff events.
The first one is sent in `UIManager.event_handler["Suspend"]` the others are directly called from the common_exit_menu. 
But the suspend action is done in `UIManager.event_handler...` for the later ones the action is done in `devicelistener`.

It is not clean that suspend actions is handled by `UIManager.event_handler` but the others through an event.

For example: If I send a `Suspend` event, the device does not suspend (only the receivers realize that), but if I send a `Reboot` or `PowerOff` event the device does the expected thing (and the receivers realize that, as the action is done with a `nextTick`).

I will look into a unification tomorrow.
But anyway I push a commit of the current work, just for information.

Reply:
The difference is probably partly that `Suspend` is an actual input event (or at least it's generated by a Power key input handler on devices where we don't have one directly), while the other two were probably mostly implemented for manual interaction via a menu (although the poweroff on long-press has been there since pretty much forever).

Anyway, yeah, definitely not against unifying that if you can do that in a non-confusing way ;).

Reply:
> It is not clean that suspend action is handled by UIManager.event_handler but the others through an event.

Some are events, others are inputs. There may or may not be unclean aspects but fundamentally it makes sense. (E.g., the input should probably generate an event and not handle it *itself* if that happens but that's another matter.)